cmd/crypto/keys.go

```diff
@@ -0,0 +1,180 @@
+// MinIO Cloud Storage, (C) 2019 MinIO, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package crypto
+
+import (
+	"bytes"
+	"crypto/tls"
+	"errors"
+
+	key "github.com/minio/keys"
+	"github.com/minio/minio/cmd/config"
+)
+
+// KeysConfig contains the configuration required
+// to initialize and connect to a keys server.
+type KeysConfig struct {
+	Enabled bool
+
+	// The keys server endpoint.
+	Endpoint string
+
+	// The path to the TLS private key used
+	// by MinIO to authenticate to the keys
+	// server during the TLS handshake (mTLS).
+	KeyFile string
+
+	// The path to the TLS certificate used
+	// by MinIO to authenticate to the keys
+	// server during the TLS handshake (mTLS).
+	//
+	// The keys server will also allow or deny
+	// access based on this certificate.
+	// In particular, the keys server will
+	// lookup the policy that corresponds to
+	// the identity in this certificate.
+	CertFile string
+
+	// Path to a file or directory containing
+	// the CA certificate(s) that issued / will
+	// issue certificates for the keys server.
+	//
+	// This is required if the TLS certificate
+	// of the keys server has not been issued
+	// (e.g. b/c it's self-signed) by a CA that
+	// MinIO trusts.
+	CAPath string
+
+	// The default key ID returned by KMS.KeyID().
+	DefaultKeyID string
+}
+
+// Verify verifies if the keys configuration is correct
+func (k KeysConfig) Verify() (err error) {
+	switch {
+	case k.Endpoint == "":
+		err = errors.New("crypto: missing keys endpoint")
```

Comment:
@aead, might want to use fallthrough and error wrapping to verify and report all configuration errors  in one shot

Reply:
We don't need to do it here, we should do it in one place when all config is handled. Each KV and subsystem we know which fields are mandatory and which are not so we can error out very early on than doing some cryptic error wrapping. 

Will send that as a separate PR